File: src/SumOfFactors.java
/**
 * Most efficient way to compute the sum of divisors of N.
 *
 * Solution:
 * Here is the most efficient which will most probably meet time constraints.
 * For every divisor i of N, there is also a corresponding divisisor N/i.
 * Thus to find all pairs of divisors, you need only to loop from 1 to the square root N.
 */
public class SumOfFactors {
    public static void main(String[] args) {
        //Test Cases
        int[] tests = {4, 10, 21, 55, 1005, 2000034};
        for (int test : tests) {
            System.out.println("The sum of the factors of " + test + " is: " + getSumOfFactors(test));
        }
    }

    private static int getSumOfFactors(int test) {
        int maxLength = (int) Math.sqrt(test);
        int sum = 1;
        for (int i = 2; i <= maxLength; i++) {
            if (test % i == 0) {
                sum += i;
                int divisor = test / i;
                if (divisor != i)
                    sum += divisor;
            }
        }
        return sum;
    }
}

File: src/kotlin_version/SumOfFactors.kt
package kotlin_version

/**
 * Most efficient way to compute the sum of divisors of N.

 * Solution:
 * Here is the most efficient which will most probably meet time constraints.
 * For every divisor i of N, there is also a corresponding divisisor N/i.
 * Thus to find all pairs of divisors, you need only to loop from 1 to the square root N.
 */
object SumOfFactors {
    @JvmStatic fun main(args: Array<String>) {
        //Test Cases
        val tests = intArrayOf(4, 10, 21, 55, 1005, 2000034)
        for (test in tests) {
            println("The sum of the factors of " + test + " is: " + getSumOfFactors(test))
        }
    }

    private fun getSumOfFactors(test: Int): Int {
        val maxLength = Math.sqrt(test.toDouble()).toInt()
        var sum = 1
        for (i in 2..maxLength) {
            if (test % i == 0) {
                sum += i
                val divisor = test / i
                if (divisor != i)
                    sum += divisor
            }
        }
        return sum
    }
}
